src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -0,0 +1,532 @@
+    if _is_exported(target):
+      exported_targets.append(ExportedTarget(target))
+    else:
+      nonexported_targets.append(target)
+  if nonexported_targets:
+    raise TargetNotExported(
+      'Cannot run setup.py on these targets, because they have no `provides=` clause: '
+      f'{", ".join(so.address.reference() for so in nonexported_targets)}')
+
+  for target in exported_targets:
+    addr = target.hydrated_target.address.reference()
+    chroot = await Get[SetupPyChroot](SetupPyChrootRequest(target))
+    provides = target.hydrated_target.adaptor.provides
+    setup_py_dir = distdir.relpath / f'{provides.name}-{provides.version}'
+    if options.values.args:
+      console.print_stderr(f'Running setup.py for {addr}')
+      setup_py_result = await Get[RunSetupPyResult](
+        RunSetupPyRequest(target, chroot, tuple(options.values.args)))
+      result_digest = setup_py_result.output
+    else:
+      # Just dump the chroot.
+      result_digest = chroot.digest
+    workspace.materialize_directory(
+      DirectoryToMaterialize(result_digest, path_prefix=str(setup_py_dir))
+    )
+    console.print_stderr(f'Wrote setup.py chroot for {addr} to {setup_py_dir}')
+
+  return SetupPy(0)
+
+
+# We write .py sources into the chroot under this dir.
+CHROOT_SOURCE_ROOT = 'src'
+
+
+SETUP_BOILERPLATE = """
+# DO NOT EDIT THIS FILE -- AUTOGENERATED BY PANTS
+# Target: {target_address_spec}
+
+from setuptools import setup
+
+setup(**{setup_kwargs_str})
+"""
+
+
+@rule
+async def run_setup_py(
+    req: RunSetupPyRequest,
+    setuptools_setup: SetuptoolsSetup,
+    python_setup: PythonSetup,
+    subprocess_encoding_environment: SubprocessEncodingEnvironment
+) -> RunSetupPyResult:
+  """Run a setup.py command on a single exported target."""
+  merged_input_files = await Get[Digest](
+    DirectoriesToMerge(directories=(
+      req.chroot.digest,
+      setuptools_setup.requirements_pex.directory_digest))
+  )
+  request = setuptools_setup.requirements_pex.create_execute_request(
+    python_setup=python_setup,
+    subprocess_encoding_environment=subprocess_encoding_environment,
+    pex_path="./setuptools.pex",
+    pex_args=('setup.py', *req.args),
+    input_files=merged_input_files,
+    output_files=('**',),
+    description=f'Run setuptools for {req.exported_target.hydrated_target.address.reference()}',
+  )
+  result = await Get[ExecuteProcessResult](ExecuteProcessRequest, request)
+  return RunSetupPyResult(result.output_directory_digest)
+
+
+@rule
+async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+  if request.py2:
+    # TODO: Implement Python 2 support.  This will involve, among other things: merging ancestor
+    # __init__.py files into the chroot, detecting packages based on the presence of __init__.py,
+    # and inspecting all __init__.py files for the namespace package incantation.
+    raise UnsupportedPythonVersion('Running setup.py commands not supported for Python 2.')
```

Comment:
To clarify, will Python 3 work when not using PEP 420? It's valid to have a Python 3-only codebase that doesn't use implicit namespace packages, e.g. the Pants repo.

Either way, totally okay with punting on this functionality. Only want to clarify what the missing functionality is and possibly update the error message.

Reply:
Yes, that should work. What won't work is python3 but with explicit namespace package incantations in `__init__.py` files. I'm not sure those are allowed in python3 anyway though. PEP420 implies that they are not.

Reply:
In other words, I think you are always "using PEP 420" in python3. The PEP allows for non-namespace packages, by having explicit `__init__.py` files. What it doesn't appear to allow is for those `__init__.py` files to create namespace packages using the python2-style.